File: dnjs/tokeniser.py
from dataclasses import dataclass
from functools import partial, lru_cache
from pathlib import Path
from typing import Dict, Iterator, Optional, Union
from string import ascii_letters, digits
import uuid


@dataclass
class Token:
    type: str
    value: str
    filepath: Optional[Path]
    pos: int
    lineno: int
    linepos: int


# you should never see this
_void_token = Token(type="VOID", value="VOID", filepath=None, pos=0, lineno=0, linepos=0)

# token types
name, string, number, template, literal = "name", "string", "number", "template", "literal"
# = => ( ) { } [ ] , : . ... ? === import from export default const `
eof = "\x03"  # end of text character
unexpected = "unexpected"
apply = "$"
many = "*"
d_name, d_many, d_brack, d_brace = "dname", "d*", "d[", "d{"
atoms = {name, string, number, template, literal, d_name}


whitespace = set(" \t\f\r")  # note no \n
_number_begin = set("-" + digits)
_number_all = set("." + digits)
_name_begin = set("_" + ascii_letters)
_name_all = set("_" + ascii_letters + digits)
_literal_values = ["null", "true", "false"]
_keyword_values = ["import", "from", "export", "default", "const"]
_punctuation_values = ["=", "(", ")", "{", "}", "[", "]", ",", ":", ".", "?", "=>", "...", "==="]
_interim_punctuation_values = ["..", "=="]

class TokenStreamEmptyError(RuntimeError):
    pass


UUID_SOURCE_MAP: Dict[uuid.UUID, str] = {}


@dataclass
class TokenStream:
    # should never raise an error, only return "unexpected" tokens
    filepath: Union[Path, uuid.UUID]
    current: Token = _void_token

    _source: Optional[str] = None
    _pos: int = 0
    _lineno: int = 1
    _linepos: int = 0
    _template_depth: int = 0

    @classmethod
    def from_source(cls, source: str):
        source_uuid = uuid.uuid4()
        UUID_SOURCE_MAP[source_uuid] = source.rstrip()
        return cls(filepath=source_uuid)

    @property
    def source(self) -> str:
        if isinstance(self.filepath, uuid.UUID):
            return UUID_SOURCE_MAP[self.filepath]
        if self._source is None:
            self._source = self.filepath.read_text().rstrip()
        return self._source

    def advance(self) -> None:
        if self.current.type == eof:
            return
        current = self._read()
        while current.type == "\n":
            current = self._read()
        self.current = current

    def __post_init__(self):
        self.advance()

    def __repr__(self) -> str:
        return f"<TokenStream file:{self.filepath}>"

    def _read(self) -> Token:
        def char() -> str:
            if self._pos == len(self.source):
                return eof
            return self.source[self._pos]

        def inc() -> None:
            self._pos += 1
            self._linepos += 1

        def inc_line() -> None:
            self._linepos = 0
            self._lineno += 1

        def gobble():
            before = char()
            inc()
            return before

        def at_comment() -> bool:
            if self._pos >= len(self.source) - 1:
                return False
            return self.source[self._pos] + self.source[self._pos + 1] == "//"

        # eat up whitespace and comments
        while char() in whitespace or at_comment():
            if at_comment():
                inc()
                inc()
                while char() not in {"\n", eof}:
                    inc()
            else:
                inc()

        make = partial(Token, pos=self._pos, filepath=self.filepath, lineno=self._lineno, linepos=self._linepos)
        t = char()
        inc()

        if t == eof:
            return make(eof, eof)

        if t == "\n":
            inc_line()
            return make("\n", "\n")

        if t == '"':
            while True:
                this_char = char()
                if this_char == eof:
                    return make(unexpected, t)
                if this_char == "\n":
                    t += gobble()
                    return make(unexpected, t)
                elif this_char == "\\":
                    gobble()
                    t += gobble()
                elif this_char == '"':
                    t += gobble()
                    return make(string, t)
                else:
                    t += gobble()

        if t == "`" or (t == "}" and self._template_depth):
            if t == "`":
                self._template_depth += 1
            while True:
                this_char = char()
                if this_char == eof:
                    return make(unexpected, t)
                if this_char == "\\":
                    gobble()
                    t += gobble()
                elif this_char == "$":
                    t += gobble()
                    if char() == "{":
                        t += gobble()
                        return make("`" if  t[0] == "`" else template, t)
                elif this_char == "`":
                    self._template_depth -= 1
                    t += gobble()
                    return make("`" if  t[0] == "`" else template, t)
                elif this_char == "\n":
                    inc_line()
                    t += gobble()
                else:
                    t += gobble()

        if t in _punctuation_values:
            if t + char() in _punctuation_values + _interim_punctuation_values:
                t += gobble()
                if t + char() in _punctuation_values:
                    t += gobble()
                    return make(t, t)
                if t in _interim_punctuation_values:
                    return make(unexpected, t)
                return make(t, t)
            return make(t, t)

        if t in _number_begin:
            seen_decimal_point = False
            while char() in _number_all:
                digit = char()
                inc()
                if digit == ".":
                    if seen_decimal_point:
                        t += digit
                        return make(unexpected, t)
                    seen_decimal_point= True
                t += digit
            return make(number, t)

        if t in _name_begin:
            while char() in _name_all:
                t += gobble()
            if t in _keyword_values:
                return make(t, t)
            if t in _literal_values:
                return make(literal, t)
            return make(name, t)

        return make(unexpected, t)

File: dnjs-go/tokeniser.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

type Token struct {
	Type  string
	Value string
	// if this starts with memory:// the source is stored in a global
	Filepath *string
	Pos      int
	Lineno   int
	Linepos  int
}

func (t Token) String() string {
	return fmt.Sprintf("<%s %s>", t.Type, t.Value)
}

// token types
var name = "name"
var str = "str"
var number = "number"
var template = "template"
var literal = "literal"

// # = => ( ) { } [ ] , : . ... ? === import from export default const `
var eof = "\x03" // end of text character
var unexpected = "unexpected"
var apply = "$"
var many = "*"
var dName = "dname"
var dMany = "d*"
var dBrack = "d["
var dBrace = "d{"
var atoms = []string{name, str, number, template, literal, dName}

var whitespace = " \t\f\r" // note no \n
var numberBegin = "-0123456789"
var numberAll = ".0123456789"
var nameBegin = "_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
var nameAll = "_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
var literalValues = []string{"null", "true", "false"}
var keywordValues = []string{"import", "from", "export", "default", "const"}
var punctuationValues = []string{"=", "=>", "(", ")", "{", "}", "[", "]", ",", ":", ".", "...", "?", "==="}
var interimPunctuationValues = []string{"..", "=="}

var globalCounter = 1
var globalSourceMap = map[string]string{}

func resetGlobalCounter() {
	globalCounter = 1
}

type TokenStream struct {
	Filepath      string
	Current       Token
	source        []rune
	pos           int
	lineno        int
	linepos       int
	templateDepth int
}

func TokenStreamFromFilepath(Filepath string) (TokenStream, error) {
	// could try do this more efficiently, but ah well
	dat, err := ioutil.ReadFile(Filepath)
	if err != nil {
		return TokenStream{}, err
	}
	tokenStream := TokenStream{
		Filepath:      Filepath,
		Current:       Token{},
		source:        []rune(strings.TrimRight(string(dat), whitespace+"\n")),
		pos:           0,
		lineno:        1,
		linepos:       0,
		templateDepth: 0,
	}
	tokenStream.Advance()
	return tokenStream, nil
}

func TokenStreamFromSource(Source string) TokenStream {
	Filepath := fmt.Sprintf("memory://%d", globalCounter)
	globalCounter += 1
	globalSourceMap[Filepath] = Source
	tokenStream := TokenStream{
		Filepath:      Filepath,
		Current:       Token{},
		source:        []rune(strings.TrimRight(Source, whitespace+"\n")),
		pos:           0,
		lineno:        1,
		linepos:       0,
		templateDepth: 0,
	}
	tokenStream.Advance()
	return tokenStream
}

func (ts *TokenStream) String() string {
	return fmt.Sprintf("<TokenStream file: %s>", ts.Filepath)
}

func (ts *TokenStream) Advance() {
	if ts.Current.Type == eof {
		return
	}
	current := ts.read()
	for {
		if current.Type != "\n" {
			break
		}
		current = ts.read()
	}
	ts.Current = current
}

func (ts *TokenStream) read() Token {
	char := func() string {
		if ts.pos == len(ts.source) {
			return eof
		}
		return string(ts.source[ts.pos])
	}
	inc := func() {
		ts.pos += 1
		ts.linepos += 1
	}
	incLine := func() {
		ts.linepos = 0
		ts.lineno += 1
	}
	gobble := func() string {
		before := char()
		inc()
		return before
	}
	atComment := func() bool {
		if ts.pos >= len(ts.source)-1 {
			return false
		}
		return string(ts.source[ts.pos])+string(ts.source[ts.pos+1]) == "//"
	}

	// eat up whitespace and comments
	for {
		if atComment() {
			inc()
			inc()
			for {
				if char() == "\n" || char() == eof {
					break
				}
				inc()
			}
		} else if strings.Contains(whitespace, char()) {
			inc()
		} else {
			break
		}
	}
	posBefore := ts.pos
	linenoBefore := ts.lineno
	lineposBefore := ts.linepos

	make := func(Type string, Value string) Token {
		return Token{
			Type:     Type,
			Value:    Value,
			Filepath: &ts.Filepath,
			Pos:      posBefore,
			Lineno:   linenoBefore,
			Linepos:  lineposBefore,
		}
	}

	t := char()
	inc()

	if t == eof {
		return make(eof, eof)
	}

	if t == "\n" {
		incLine()
		return make("\n", "\n")
	}

	if t == "\"" {
		for {
			thisChar := char()
			if thisChar == eof {
				return make(unexpected, t)
			}
			if thisChar == "\n" {
				t += gobble()
				return make(unexpected, t)
			}
			if thisChar == "\\" {
				gobble()
				t += gobble()
			} else if thisChar == "\"" {
				t += gobble()
				return make(str, t)
			} else {
				t += gobble()
			}
		}
	}

	if t == "`" || (t == "}" && ts.templateDepth > 0) {
		if t == "`" {
			ts.templateDepth += 1
		}
		for {
			thisChar := char()
			if thisChar == eof {
				return make(unexpected, t)
			}
			if thisChar == "\\" {
				gobble()
				t += gobble()
			} else if thisChar == "$" {
				t += gobble()
				if char() == "{" {
					t += gobble()
					if ([]rune(t))[0] == '`' {
						return make("`", t)
					}
					return make(template, t)
				}
			} else if thisChar == "`" {
				ts.templateDepth -= 1
				t += gobble()

				if ([]rune(t))[0] == '`' {
					return make("`", t)
				}
				return make(template, t)
			} else if thisChar == "\n" {
				incLine()
				t += gobble()
			} else {
				t += gobble()
			}
		}
	}

	if containsString(punctuationValues, t) {
		allPunctuationValues := append(punctuationValues, interimPunctuationValues...)
		if containsString(allPunctuationValues, t+char()) {
			t += gobble()
			if containsString(punctuationValues, t+char()) {
				t += gobble()
				return make(t, t)
			}
			if containsString(interimPunctuationValues, t) {
				return make(unexpected, t)
			}
			return make(t, t)
		}
		return make(t, t)
	}

	if strings.Contains(numberBegin, t) {
		seenDecimalPoint := false
		for {
			if !strings.Contains(numberAll, char()) {
				break
			}
			digit := char()
			inc()
			if digit == "." {
				if seenDecimalPoint {
					t += digit
					return make(unexpected, t)
				}
				seenDecimalPoint = true
			}
			t += digit
		}
		return make(number, t)
	}

	if strings.Contains(nameBegin, t) {
		for {
			if !strings.Contains(nameAll, char()) {
				break
			}
			t += gobble()
		}
		if containsString(keywordValues, t) {
			return make(t, t)
		}
		if containsString(literalValues, t) {
			return make(literal, t)
		}
		return make(name, t)
	}

	return make(unexpected, t)
}
